pulp_docker/tests/functional/api/test_pull_content.py

```diff
@@ -0,0 +1,215 @@
+                (cls.client.delete, cls.publisher['_href'])
+            )
+
+            # Step 5.
+            cls.publication = publish(cls.cfg, cls.publisher, cls.repo)
+            cls.teardown_cleanups.append(
+                (cls.client.delete, cls.publication['_href'])
+            )
+
+            # Step 6.
+            cls.distribution = cls.client.post(
+                DOCKER_DISTRIBUTION_PATH,
+                gen_distribution(publication=cls.publication['_href'])
+            )
+            cls.teardown_cleanups.append(
+                (cls.client.delete, cls.distribution['_href'])
+            )
+
+            # remove callback if everything goes well
+            stack.pop_all()
+
+    @classmethod
+    def tearDownClass(cls):
+        """Clean class-wide variable."""
+        for cleanup_function, args in reversed(cls.teardown_cleanups):
+            cleanup_function(args)
+
+    def test_api_returns_same_checksum(self):
+        """Verify that pulp serves image with the same checksum of remote.
+
+        1. Call pulp repository API and get the content_summary for repo.
+        2. Call dockerhub API and get blobsums for synced image.
+        3. Compare the checksums.
+        """
+        # Get local checksums for content synced from remote registy
+        checksums = [
+            content['digest'] for content
+            in get_content(self.repo)[DOCKER_CONTENT_NAME]
+        ]
+
+        # Assert that at least one layer is synced from remote:latest
+        # and the checksum matched with remote
+        self.assertTrue(
+            any(
+                [
+                    result['blobSum'] in checksums
+                    for result in get_docker_hub_remote_blobsums()
+                ]
+            ),
+            'Cannot find a matching layer on remote registry.'
+        )
+
+    def test_pull_image(self):
+        """Verify that a client can pull the image from Pulp.
+
+        1. Using the RegistryClient pull the image from Pulp.
+        2. Pull the same image from remote registry.
+        3. Verify both images has the same checksum.
+        4. Ensure image is deleted after the test.
+        """
+        registry = cli.RegistryClient(self.cfg)
+        registry.raise_if_unsupported(
+            unittest.SkipTest, 'Test requires podman/docker'
+        )
+
+        local_url = urljoin(
+            self.cfg.get_content_host_base_url(),
+            self.distribution['base_path']
+        )
+
+        registry.pull(local_url)
+        self.teardown_cleanups.append((registry.rmi, local_url))
+        local_image = registry.inspect(local_url)
+
+        registry.pull(DOCKER_UPSTREAM_NAME)
+        self.teardown_cleanups.append((registry.rmi, DOCKER_UPSTREAM_NAME))
+        remote_image = registry.inspect(DOCKER_UPSTREAM_NAME)
+
+        self.assertEqual(
+            local_image[0]['Digest'],
+            remote_image[0]['Digest']
+        )
+
+    def test_pull_image_with_tag(self):
```

Comment:
i suggest to also add pull image by digest

Reply:
Will also land in a new PR later